File: samples/sample-app-storage/src/main/java/com/azure/android/storage/sample/UploadFileActivity.java
package com.azure.android.storage.sample;

import android.app.Activity;
import android.net.Uri;
import android.os.Bundle;
import android.util.Log;
import android.widget.ProgressBar;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.LifecycleOwner;

import com.azure.android.identity.TokenRequestObservable;
import com.azure.android.identity.TokenRequestObservableAuthInterceptor;
import com.azure.android.identity.TokenRequestObserver;
import com.azure.android.identity.TokenResponseCallback;
import com.azure.android.storage.blob.StorageBlobClient;
import com.azure.android.storage.blob.transfer.TransferClient;
import com.azure.android.storage.sample.config.StorageConfiguration;
import com.microsoft.identity.client.IMultipleAccountPublicClientApplication;
import com.microsoft.identity.client.PublicClientApplication;
import com.microsoft.identity.client.exception.MsalException;

import java.io.File;
import java.util.Collections;
import java.util.List;

import javax.inject.Inject;

import static com.azure.android.storage.sample.Constants.FILE_URI_EXTRA;

public class UploadFileActivity extends AppCompatActivity {
    private static final String TAG = UploadFileActivity.class.getSimpleName();

    private ProgressBar progressBar;
    private StorageConfiguration storageConfiguration;

    // Singleton StorageBlobClient that will be created by Dagger. The singleton object is shared across various
    // activities in the application.
    @Inject
    StorageBlobClient storageBlobClient;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_uploadfile);

        // Request Dagger to get singleton StorageBlobClient and initialize this.storageBlobClient
        ((MainApplication) getApplication()).getAppComponent().inject(this);

        this.progressBar = findViewById(R.id.progressBar);
        this.storageConfiguration = StorageConfiguration.create(getApplicationContext());

        // Set up Login
        final List<String> blobEndpointScopes = Collections.singletonList(storageBlobClient.getBlobServiceUrl() + ".default");
        TokenRequestObservableAuthInterceptor authInterceptor =
            new TokenRequestObservableAuthInterceptor(blobEndpointScopes);

        TokenRequestObservable tokenRequestObservable = authInterceptor.getTokenRequestObservable();
        LifecycleOwner lifecycleOwner = this;

        PublicClientApplication.createMultipleAccountPublicClientApplication(
            this.getApplicationContext(),
            R.raw.authorization_configuration,
            new PublicClientApplication.IMultipleAccountApplicationCreatedListener() {
                @Override
                public void onCreated(IMultipleAccountPublicClientApplication application) {
                    tokenRequestObservable.observe(lifecycleOwner, new TokenRequestObserver() {
                        @Override
                        public void onTokenRequest(String[] scopes, TokenResponseCallback callback) {
                            MsalClient.signIn(application, getActivity(), scopes, callback);
                        }
                    });
                }

                @Override
                public void onError(MsalException exception) {
                    Log.e(TAG, "Exception found when trying to sign in.", exception);
                }
            });

        // Create a new StorageBlobClient from the existing client with different base URL and credentials but sharing
        // the underlying OkHttp Client.
        storageBlobClient = storageBlobClient
            .newBuilder()
            .setBlobServiceUrl(storageConfiguration.getBlobServiceUrl())
            .setCredentialInterceptor(authInterceptor)
            .build();
    }

    @Override
    protected void onResume() {
        super.onResume();

        Uri fileUri = getIntent().getParcelableExtra(FILE_URI_EXTRA);
        //String contentType = this.getContentResolver().getType(fileUri);
        String filePath = PathUtil.getPath(this, fileUri);
        File fileToUpload = new File(filePath);
        int fileSize = (int) fileToUpload.length();

        this.progressBar.setMax(fileSize);

        final String containerName = storageConfiguration.getContainerName();
        final String blobName = fileToUpload.getName();

        Log.d("Upload file", "File path: " + filePath);
        Log.d("Upload file", "Blob name: " + blobName);
        Log.d("Upload file", "File size: " + fileSize);

        try {
            TransferClient transferClient = new TransferClient.Builder(getApplicationContext())
                .addStorageBlobClient(Constants.STORAGE_BLOB_CLIENT_ID, storageBlobClient)
                .setRequiredNetworkType(androidx.work.NetworkType.CONNECTED)
                .build();

            transferClient.upload(Constants.STORAGE_BLOB_CLIENT_ID, containerName, blobName, fileToUpload)
                .observe(this, new TransferObserver() {
                    @Override
                    public void onStart(long transferId) {
                        Log.i(TAG, "onStart()");
                    }

                    public void onProgress(long transferId, long totalBytes, long bytesTransferred) {
                        Log.i(TAG, "onProgress(" + totalBytes + ", " + bytesTransferred + ")");

                        progressBar.setProgress((int) bytesTransferred);
                    }

                    public void onSystemPaused(long transferId) {
                        Log.i(TAG, "onSystemPaused()");
                    }

                    public void onResume(long transferId) {
                        Log.i(TAG, "onResumed()");
                    }

                    public void onComplete(long transferId) {
                        Log.i(TAG, "onCompleted()");

                        progressBar.setProgress(fileSize);

                        Toast.makeText(getApplicationContext(), "Upload complete", Toast.LENGTH_SHORT).show();
                    }

                    public void onError(long transferId, String errorMessage) {
                        Log.i(TAG, "onError() -> : " + errorMessage);
                    }
                });
        } catch (Exception ex) {
            Log.e(TAG, "Upload submit failed: ", ex);
        }
    }

    private Activity getActivity() {
        return this;
    }
}

File: samples/sample-app-storage-kotlin/src/main/java/com/azure/android/storage/sample/kotlin/UploadFileActivity.kt
package com.azure.android.storage.sample.kotlin

import android.app.Activity
import android.net.Uri
import android.os.Bundle
import android.provider.OpenableColumns
import android.util.Log
import android.widget.ProgressBar
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.LifecycleOwner
import androidx.work.NetworkType
import com.azure.android.core.credential.TokenRequestObservable
import com.azure.android.core.credential.TokenRequestObservableAuthInterceptor
import com.azure.android.core.credential.TokenRequestObserver
import com.azure.android.core.credential.TokenResponseCallback
import com.azure.android.storage.blob.StorageBlobAsyncClient
import com.azure.android.storage.blob.transfer.TransferClient
import com.azure.android.storage.sample.kotlin.config.StorageConfiguration
import com.microsoft.identity.client.IMultipleAccountPublicClientApplication
import com.microsoft.identity.client.IPublicClientApplication.IMultipleAccountApplicationCreatedListener
import com.microsoft.identity.client.PublicClientApplication
import com.microsoft.identity.client.exception.MsalException

class UploadFileActivity : AppCompatActivity() {
    private lateinit var progressBar: ProgressBar
    private lateinit var storageConfiguration: StorageConfiguration
    private lateinit var storageBlobAsyncClient: StorageBlobAsyncClient

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_uploadfile)
        progressBar = findViewById(R.id.progressBar)

        val initStorageBlobAsyncClient: StorageBlobAsyncClient = StorageBlobAsyncClient
            .Builder("upload-file-activity")
            .setBlobServiceUrl(storageConfiguration.blobServiceUrl)
            .build()

        // Set up Login
        storageConfiguration = StorageConfiguration.create(applicationContext)
        val blobEndpointScopes = listOf(initStorageBlobAsyncClient.getBlobServiceUrl() + ".default")
        val authInterceptor = TokenRequestObservableAuthInterceptor(blobEndpointScopes)
        val tokenRequestObservable: TokenRequestObservable = authInterceptor.tokenRequestObservable
        val lifecycleOwner: LifecycleOwner = this
        PublicClientApplication.createMultipleAccountPublicClientApplication(
            this.applicationContext,
            R.raw.authorization_configuration,
            object : IMultipleAccountApplicationCreatedListener {
                override fun onCreated(application: IMultipleAccountPublicClientApplication) {
                    tokenRequestObservable.observe(lifecycleOwner, object : TokenRequestObserver() {
                        override fun onTokenRequest(scopes: Array<String>, callback: TokenResponseCallback) {
                            MsalClient.signIn(application, activity, scopes, callback)
                        }
                    })
                }

                override fun onError(exception: MsalException) {
                    Log.e(TAG, "Exception found when trying to sign in.", exception)
                }
            })

        // Create a new StorageBlobClient from the existing client with different base URL and credentials but sharing
        // the underlying OkHttp Client.
        storageBlobAsyncClient = storageBlobAsyncClient
            .newBuilder("com.azure.android.storage.sample.upload")
            .setBlobServiceUrl(storageConfiguration.blobServiceUrl)
            .setCredentialInterceptor(authInterceptor)
            .setTransferRequiredNetworkType(NetworkType.CONNECTED)
            .build()
    }

    override fun onResume() {
        super.onResume()
        val fileUri : Uri? = intent.getParcelableExtra<Uri>(Constants.FILE_URI_EXTRA)
        // Use content resolver to get file size and name.
        // https://developer.android.com/training/secure-file-sharing/retrieve-info
        val cursor = contentResolver.query(fileUri!!, null, null, null, null)
        val sizeIndex = cursor!!.getColumnIndex(OpenableColumns.SIZE)
        val nameIndex = cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME)
        cursor.moveToFirst()
        val fileSize = cursor.getLong(sizeIndex)
        val blobName = cursor.getString(nameIndex)
        val containerName: String? = storageConfiguration.containerName
        progressBar.max = fileSize.toInt()
        Log.d("Upload Content", "Content Uri: $fileUri")
        Log.d("Upload Content", "Blob name: $blobName")
        Log.d("Upload Content", "File size: $fileSize")
        try {
            storageBlobAsyncClient.upload(applicationContext, containerName, blobName, false, fileUri)
                .observe(this, object : TransferObserver {
                    override fun onStart(transferId: Long) {
                        Log.i(TAG, "onStart()")
                    }

                    override fun onProgress(transferId: Long, totalBytes: Long, bytesTransferred: Long) {
                        Log.i(TAG, "onProgress($totalBytes, $bytesTransferred)")
                        progressBar.progress = bytesTransferred.toInt()
                    }

                    override fun onSystemPaused(transferId: Long) {
                        Log.i(TAG, "onSystemPaused()")
                    }

                    override fun onResume(transferId: Long) {
                        Log.i(TAG, "onResumed()")
                    }

                    override fun onComplete(transferId: Long) {
                        Log.i(TAG, "onCompleted()")
                        progressBar.progress = fileSize.toInt()
                        Toast.makeText(applicationContext, "Upload complete", Toast.LENGTH_SHORT).show()
                    }

                    override fun onError(transferId: Long, errorMessage: String?) {
                        Log.i(TAG, "onError() -> : $errorMessage")
                    }
                })
        } catch (ex: Exception) {
            Log.e(TAG, "Upload submit failed: ", ex)
        }
    }

    private val activity: Activity
        get() = this

    companion object {
        private val TAG = UploadFileActivity::class.java.simpleName
    }
}

